### PR TITLE
Refactor placement interfaces

### PR DESCRIPTION
**Pull request by cw9**
Part of the Service Discovery interface change.

No logic change, all meaningful changes are in types.go

@xichen2020 @robskillington

### CODE REVIEW

File: services/types.go [hunk 1]
```diff
@@ -69,49 +90,137 @@ type ServiceSharding interface {
...
+
+	// ReplaceInstance picks instances from the candidate list to replace an instance in current placement
+	ReplaceInstance(leavingInstance PlacementInstance, candidates []PlacementInstance, popts PlacementOptions) (ServicePlacement, error)
+
+	// Placement gets the persisted placement for service
+	Placement() (ServicePlacement, error)
+}
+
+// PlacementOptions is the interface for placement options
+type PlacementOptions interface {
```
Comment (xichen2020) [inline services/types.go]: Move the options to the service constructor

File: services/types.go [hunk 2]
```diff
@@ -69,49 +90,137 @@ type ServiceSharding interface {
...
+	ReplicaFactor() int
+
+	// Shards returns all the unique shard ids for a replica
+	Shards() []uint32
+
+	// ShardsLen returns the number of shards in a replica
+	NumShards() int
+
+	// Validate checks if the ServicePlacement is valid
+	Validate() error
```
Comment (xichen2020) [inline services/types.go]: Check if `Validate` and `Copy` are only used inside the placement service and if so, move them inside `PlacementService`

File: services/types.go [hunk 3]
```diff
@@ -69,49 +90,137 @@ type ServiceSharding interface {
...
+	String() string                             // String is for debugging
+	ID() string                                 // ID is the id of the instance
+	SetID(id string) PlacementInstance          // SetID sets the id of the instance
+	Rack() string                               // Rack is the rack of the instance
+	SetRack(r string) PlacementInstance         // SetRack sets the rack of the instance
+	Zone() string                               // Zone is the zone of the instance
+	SetZone(z string) PlacementInstance         // SetZone sets the zone of the instance
+	Weight() uint32                             // Weight is the weight of the instance
+	SetWeight(w uint32) PlacementInstance       // SetWeight sets the weight of the instance
+	IP() string                                 // IP is the ip of the instance
```
Comment (xichen2020) [inline services/types.go]: Merge `IP` and `Port` into `Endpoint`

File: services/types.go (hunk 1 above)
Comment (cw9) [inline services/types.go]: done

File: services/services.go [hunk 4]
```diff
@@ -76,21 +76,21 @@ func (s *serviceSharding) SetNumShards(n int) ServiceSharding { s.numShards = n;
...
-func (i *serviceInstance) Endpoint() string                         { return i.endpoint }
-func (i *serviceInstance) Shards() shard.Shards                     { return i.shards }
-func (i *serviceInstance) SetService(s string) ServiceInstance      { i.service = s; return i }
-func (i *serviceInstance) SetID(id string) ServiceInstance          { i.id = id; return i }
-func (i *serviceInstance) SetZone(z string) ServiceInstance         { i.zone = z; return i }
+func (i *serviceInstance) ServiceID() ServiceID { return i.service }
+func (i *serviceInstance) InstanceID() string   { return i.id }
+func (i *serviceInstance) Endpoint() string     { return i.endpoint }
+func (i *serviceInstance) Shards() shard.Shards { return i.shards }
+func (i *serviceInstance) SetServiceID(service ServiceID) ServiceInstance {
```
Comment (xichen2020) [inline services/services.go]: Can we put a blank line between this method and single-line methods above and below?

**Pull request merged**